tests/src/tracing/eventpipe/common/StreamProxy.cs

```diff
@@ -0,0 +1,112 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.IO;
+
+namespace Tracing.Tests.Common
+{
+    // This Stream implementation takes one stream
+    // and proxies the Stream API to it while
+    // saving any read bytes to an internal stream.
+    // Should an error occur, the internal stream
+    // is dumped to disk for reproducing the error.
+    public class StreamProxy : Stream
+    {
+        private Stream ProxiedStream { get; }
+        private MemoryStream InternalStream => new MemoryStream();
+        public override bool CanRead => ProxiedStream.CanRead;
+
+        public override bool CanSeek => ProxiedStream.CanSeek;
+
+        public override bool CanWrite => ProxiedStream.CanWrite;
+
+        public override long Length => ProxiedStream.Length;
+
+        public override long Position { get => ProxiedStream.Position; set => ProxiedStream.Position = value; }
+
+        public StreamProxy(Stream streamToProxy)
+        {
+            ProxiedStream = streamToProxy;
+        }
+
+        public override void Flush() => ProxiedStream.Flush();
+
+        // Read the actual desired amount of bytes into an empty buffer
+        // copy those bytes into an internal MemoryStream THEN
+        // forward those bytes to the caller. If the caller
+        // would have thrown an exception with its Read, copy
+        // the bytes to the internal MemoryStream and then throw
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null || offset < 0 || count < 0)
+                throw new ArgumentException("Invalid input into Read");
+
+            byte[] localBuffer = ArrayPool<byte>.Shared.Rent(count);
+            var readCount = ProxiedStream.Read(localBuffer, 0, count);
+            if (readCount == 0)
+                return readCount;
+
+            InternalStream.Write(localBuffer, 0, readCount);
+
+            if (buffer.Length - offset < count)
```

Comment:
This error condition is fairly specific so you might be able to do code review on the StreamReader in TraceEvent and figure out the likely bug without having an exact repro input. (Not that I object to any of this additional test telemetry)